torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,375 @@
+    manager :class:``cached`` and the decorator :class: ``cached_method``
+    """
+
+    def evaluate(self, tensor=None):
+        r""" It evaluates the current chain of parametrizations on a given tensor.
+
+        If the tensor is None, and the parametrization has been registered on a module,
+        it evaluates the parametrizaiton on the tensor on which it was registered
+
+        Args:
+            tensor (torch.Tensor): the at which to evaluate the parametrization chain.
+            Default: None
+        """
+        # TODO Write docs
+        if self.is_chained():
+            tensor = self.orig_param.evaluate(tensor)
+        else:
+            if tensor is None:
+                try:
+                    tensor = self.orig
+                except AttributeError:
+                    raise ValueError("Cannot evaluate a Parametrization that has not "
+                                     "been registered without a tensor.")
+        return self(tensor)
+
+    def original(self):
+        # TODO Write docs
+        try:
+            return self.last_parametrization().orig
+        except AttributeError:
+            raise ValueError("The parametrization is not registered to a tensor.")
+
+    def last_parametrization(self):
+        # TODO Write docs
+        last = self
+        while last.is_chained():
+            last = last.orig_param
+        return last
+
+    def chain(self, parametrization):
+        r""" It modifies the current parametrization, chaining a new one to it.
+        We have that ``p2.evaluate(p1.evaluate(t)) == p2.chain(p1).evaluate(t)``
+
+        The original parametrization is modified after chain is applied to it
+
+        Args:
+            parametrization (nn.Parametrization): the parametrization to be chained
+        Returns:
+            Module: The original module with the parametrization module appended to it
+        """
+        if not isinstance(parametrization, Parametrization):
+            raise ValueError("The object '{}' is not a Parametrization"
+                             .format(parametrization))
+
+        if self.is_registered() and parametrization.is_registered():
+            raise ValueError("Both parametrizations are registered to a tensor. "
+                             "The chaining is ambiguous.")
+        last = self.last_parametrization()
+        is_parametrized = hasattr(last, "orig")
+        if is_parametrized:
+            # We move the orig tensor to the new last parametrization
+            orig = last.orig
+            delattr(last, "orig")
+
+        _set_parametrization(last, parametrization, "orig", set_cache=False)
+        if is_parametrized:
+            new_last = parametrization.last_parametrization()
+            if isinstance(orig, Parameter):
+                new_last.register_parameter("orig", orig)
+            else:
+                new_last.register_buffer("orig", orig)
+        return self
```

Comment:
This seems quite complicated, and I think it could be improved once you make `Parametrization` objects less stateful. You should be able to construct multiple of those, connect them into a container akin to `nn.Sequential` (maybe using this `>>` operator). Then you get this:
```py
class ParametrizationChain(nn.Module):
  def __init__(self, *parametrizations):
    self.parametrizations = parametrizations

  def register(self, tensor):
    self.parametrizations[0].register(tensor)

  def evaluate(self, tensor=None):
    for p in parametrizations:
      tensor = p.evaluate(tensor)
```
Note that I've isolated the `register` function to make sure you don't have to much around too much with the `orig` attribute in `register_parametrization`. It should defer setting up of the state of the parametrization to the actual object.

Reply:
Answered above.